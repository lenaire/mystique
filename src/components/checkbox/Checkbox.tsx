import { useId, forwardRef, useCallback, ChangeEvent, useState } from "react";
import { Icon } from "$cmp/icon/Icon";
import styled from "styled-components";
import { CmpProps } from "$types";

const Container = styled.div``;

const Label = styled.label`
  display: flex;
  align-items: center;
  cursor: pointer;
`;

const Input = styled.input`
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(50%);
  height: 1px;
  width: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
`;

const Box = styled.span<{ $error: boolean; $checked: boolean }>`
  height: 20px;
  width: 20px;
  background-color: ${(props): string =>
    props.$checked ? props.theme.palette.primary.main : props.theme.palette.common.white};
  border: 1px solid
    ${(props): string => {
      if (props.$error) return props.theme.palette.error.main;

      if (props.$checked) return props.theme.palette.primary.main;

      return props.theme.palette.grey[300];
    }};
  border-radius: 4px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
`;

const Check = styled(Icon)`
  fill: ${(props): string => props.theme.palette.common.white};
  height: 100%;
  width: 100%;
`;

const Text = styled.span`
  padding-left: 8px;
`;

export interface CheckboxProps extends CmpProps {
  id?: string;
  label?: string;
  ariaLabel?: string;
  defaultValue?: boolean;
  onChange: (val: boolean) => void;
  onBlur?: (val: boolean) => void;
  error?: boolean;
}

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {
      id,
      label,
      onChange,
      onBlur,
      ariaLabel,
      className,
      defaultValue = false,
      error = false,
      ...rest
    },
    ref
  ) => {
    const [checked, setChecked] = useState(defaultValue);
    const inputId = useId();
    const elemId = id || inputId;

    const handleChange = useCallback(
      (event: ChangeEvent<HTMLInputElement>) => {
        onChange(event.target.checked);
        onBlur?.(event.target.checked);
        setChecked(!checked);
      },
      [onChange, onBlur, checked]
    );

    return (
      <Container className={className} {...rest}>
        <Label aria-label={ariaLabel} htmlFor={elemId}>
          <Input
            id={elemId}
            type="checkbox"
            checked={checked}
            onChange={handleChange}
            data-testid="toggle-input"
            ref={ref}
          />
          {/* TODO: Typography component */}
          <Box $checked={checked} $error={!!error}>
            <Check icon="check" />
          </Box>
          {label && <Text>{label}</Text>}
        </Label>
      </Container>
    );
  }
);
