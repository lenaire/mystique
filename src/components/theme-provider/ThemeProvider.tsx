import { FC } from "react";
import {
  DefaultTheme, // This is overridden in types/theme.d.ts
  ThemeProvider as StyledComponentsThemeProvider,
  ThemeProviderProps as StyledComponentsThemeProviderProps,
} from "styled-components";

import { defaultTheme } from "$cmp/theme-provider/theme";
import { ResetGlobalCSS } from "$cmp/theme-provider/ResetGlobalCSS";

export type ThemeProviderProps = Partial<StyledComponentsThemeProviderProps<DefaultTheme>>;

export const ThemeProvider: FC<ThemeProviderProps> = ({ children, theme = defaultTheme }) => {
  return (
    <StyledComponentsThemeProvider theme={theme}>
      <ResetGlobalCSS />
      {children}
    </StyledComponentsThemeProvider>
  );
};
