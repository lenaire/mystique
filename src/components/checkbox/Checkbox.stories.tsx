import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Checkbox } from "./Checkbox";

export default {
  title: "Checkbox",
  component: Checkbox,
  argTypes: {
    className: { control: { disable: true } },
    onChange: { action: "toggled!" },
  },
} as ComponentMeta<typeof Checkbox>;

const Template: ComponentStory<typeof Checkbox> = (args) => <Checkbox {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  label: "This is a toggle",
  defaultValue: true,
};
