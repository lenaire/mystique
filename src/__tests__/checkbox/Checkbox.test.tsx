import renderer from "react-test-renderer";
import { ThemeProvider } from "$cmp/theme-provider/ThemeProvider";
import { Checkbox, CheckboxProps } from "$cmp/checkbox/Checkbox";

describe("<Checkbox />", () => {
  it("renders correctly", () => {
    const props: CheckboxProps = {
      onChange: jest.fn()
    };

    const tree = renderer
      .create(
        <ThemeProvider>
          <Checkbox {...props} />
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
