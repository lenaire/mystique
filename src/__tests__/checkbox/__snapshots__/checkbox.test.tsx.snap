// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> renders correctly 1`] = `
<div
  className="sc-gKPRtg"
>
  <label
    className="sc-iBYQkv btqRMo"
    htmlFor=":r0:"
  >
    <input
      checked={false}
      className="sc-ftTHYK qoTa-D"
      data-testid="toggle-input"
      id=":r0:"
      onChange={[Function]}
      type="checkbox"
    />
    <span
      className="sc-pyfCe ebpNrf"
    >
      <div
        className="sc-gswNZR FbElD sc-jrcTuL ezIwCV"
        title="A(n) check icon"
      />
    </span>
  </label>
</div>
`;
